src/transport/raw/TCP.cpp

```diff
@@ -354,11 +354,40 @@ CHIP_ERROR TCPBase::ProcessReceivedBuffer(Inet::TCPEndPoint * endPoint, const Pe
             continue;
         }
 
-        // Buffer is incomplete if we reach this point
+        // If we reach this point, the head buffer does not contain a complete message.
+        // We may or may not have a complete message in a buffer chain.
         if (buffer->HasChainedBuffer())
         {
-            buffer->CompactHead();
-            continue;
+            if (messageSize > System::PacketBuffer::kMaxSizeWithoutReserve)
+            {
+                // This message will not fit in a maximum-size buffer.
+                if (messageSize <= buffer->TotalLength())
+                {
+                    // We have the oversize message in the current buffer chain.
+                    // Since it's oversize, it is definitely not a valid CHIP message, so we angrily toss it aside.
+                    buffer->ConsumeHead(messageSize);
+                    err = endPoint->AckReceive(messageSize);
```

Comment:
And I would guess that this too should be `messageSize + kPacketSizeBytes`, as should the existing `AckReceive` call we have in the `ContainsCompleteMessage` case.....  That one is a pre-existing bug.